File: uva online judge/10003 - Cutting Sticks.c
#include<stdio.h>
#include<limits.h>
#define MAX_SIZE 51
#define INF INT_MAX

int cuts[MAX_SIZE], dp[MAX_SIZE][MAX_SIZE];

int read_test_case(int cuts[]){
    int l, n, i;
    scanf("%d", &l);
    if(l == 0)
        return 0;
    scanf("%d", &n);
    for(i = 1; i <= n; i++)
        scanf("%d", &cuts[i]);
    cuts[n + 1] = l;
    return n + 2;
}

int cutting_sticks_dp(int cuts[], int i, int j, int dp[][MAX_SIZE]){
    int k, value, min_value = INF;
    if(j - i == 1)
        return 0;
    if(dp[i][j] == 0){
        for(k = i + 1; k < j; k++){
            value = cutting_sticks_dp(cuts, i, k, dp) + cutting_sticks_dp(cuts, k, j, dp);
            if(value < min_value)
                min_value = value;
        }
        dp[i][j] = min_value + cuts[j] - cuts[i];
    }
    return dp[i][j];
}

int main(){
    int i, j, n;
    while((n = read_test_case(cuts)) > 0){
        for(i = 0; i < n; i++)
            for(j = 0; j < n; j++)
                dp[i][j] = 0;
        printf("The minimum cutting is %d.\n", cutting_sticks_dp(cuts, 0, n - 1, dp));
    }
    return 0;
}

File: uva online judge/10003 - Cutting Sticks.py
from math import inf

def read_test_cases():
    while True:
        cuts = [0]
        l = int(input())
        if l == 0:
            break
        input()
        cuts.extend(int(i) for i in input().split())
        cuts.append(l)
        yield cuts

def cutting_sticks_dp(cuts, dp):
    n = len(cuts)
    for i in range(0, n - 1):
        dp[i][i + 1] = 0
    for l in range(2, n):
        for i in range(0, n - l):
            min_value = inf
            for k in range(i + 1, i + l):
                value = dp[i][k] + dp[k][i + l]
                if value < min_value:
                    min_value = value
            dp[i][i + l] = min_value + cuts[i + l] - cuts[i]
    return dp[0][n - 1]

if __name__ == '__main__':
    max_size = 51
    dp = [[None] * max_size for i in range(max_size)]
    for cuts in read_test_cases():
        print('The minimum cutting is {}.'.format(cutting_sticks_dp(cuts, dp)))
